README.rst
=============
WS2812_driver
=============


.. image:: https://img.shields.io/pypi/v/ws2812_driver.svg
        :target: https://pypi.python.org/pypi/ws2812_driver

.. image:: https://img.shields.io/travis/phephik/ws2812_driver.svg
        :target: https://travis-ci.org/phephik/ws2812_driver

.. image:: https://readthedocs.org/projects/ws2812-driver/badge/?version=latest
        :target: https://ws2812-driver.readthedocs.io/en/latest/?badge=latest
        :alt: Documentation Status

.. image:: https://pyup.io/repos/github/phephik/ws2812_driver/shield.svg
     :target: https://pyup.io/repos/github/phephik/ws2812_driver/
     :alt: Updates

Python driver for controlling  WS2812(B) or any comparable RGB adressable LED strip over I2S bus.


* Free software: BSD license
* Documentation: https://ws2812-driver.readthedocs.io.

Overview
========
The repository includes a python code needed to build an LED strip scale thermometer:

This code for visualization includes functions:

* scale thermometer
* demo trail


Prerequisites
=============

This project is based on I2S bus, so use DOUT pin of I2S bus on your device.
Before you run code, you should make sure that you work with:

* Linux embedded platform (recommended Armbian 5.30 stable Ubuntu 16.04.2)
* ws2812b addressable led strip (of any size)
* 5V power supply (recommended)

Python Dependencies
===================

Compatible with Python 3.5 (including pyalsaaudio package)

* Tested on Debian-based Armbian (Armbian 5.30 stable Ubuntu 16.04.2).
* Linux kernel version at least 3.4.113

Installing dependencies
-----------------------

You will need to set up a I2S device with following commands:

.. code-block:: bash

    cd /boot
    sudo bin2fex script.bin script.fex
    sudo nano script.fex

In next step change following attributes in script:

.. code-block:: bash

    [twi1]
    twi_used = 0

    [pcm0]
    daudio_used = 1
    sample_resolution = 16
    slot_width_select = 16
    pcm_lrck_period = 16
    slot_width = 16

Save your changes and convert file again:

.. code-block:: bash

    sudo fex2bin script.fex script.bin

Reboot your device. After reboot try command:

.. code-block:: bash

    aplay -l

It looks like:

.. code-block:: bash

    **** List of PLAYBACK Hardware Devices ****
    card 0: audiocodec [audiocodec], device 0: SUNXI-CODEC sndcodec-0 []
    Subdevices: 1/1
    Subdevice #0: subdevice #0
    card 1: snddaudio [snddaudio], device 0: SUNXI-TDM0 snddaudio-0 []
    Subdevices: 1/1
    Subdevice #0: subdevice #0
    card 2: sndhdmi [sndhdmi], device 0: SUNXI-HDMIAUDIO sndhdmi-0 []
    Subdevices: 1/1
    Subdevice #0: subdevice #0

"card 1" is the one we want.

Demo trail
==========
* Function show_animation will display fast moving trail.
* You can change speed of trail (0-100)



Scale thermometer
=================

* Function meas_temperature sets up led string as a thermometer or scale display.
* As arguments you have to define number of leds, minimum, maximum, temperature.


.. code-block:: bash

    def meas_temperature(num_leds, min_temp, max_temp, temp, bright=100, heatmap=None, reverse=None)

* Optional arguments are brightness, your own-defined heatmap and reverse mode


.. code-block:: bash

        heatmap = [(0,   100,  (0, 0, 256)),
                   (400, 500,  (0, 256, 0)),
                   (900, 1000, (256, 0, 0))]

* You can define heatmap with fixed limits for each color e.g. (0, 150, (10, 256, 20). Coulours among are set transiently.


Credits
=======

This package was created with Cookiecutter_ and the `audreyr/cookiecutter-pypackage`_ project template.

.. _Cookiecutter: https://github.com/audreyr/cookiecutter
.. _`audreyr/cookiecutter-pypackage`: https://github.com/audreyr/cookiecutter-pypackage


HISTORY.rst
=======
History
=======

0.1.0 (2017-08-17)
------------------

* First release on PyPI.


tests/__init__.py
# -*- coding: utf-8 -*-

"""Unit test package for ws2812_driver."""


tests/test_ws2812_driver.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

"""Tests for `ws2812_driver` package."""


import unittest
import numpy as np
import time


from ws2812_driver import Ws2812Driver

class TestWs2812_driver(unittest.TestCase):
    """Tests for `ws2812_driver` package."""

    @classmethod
    def setUpClass(cls):
        """Set up test fixtures, if any."""
        cls.pasek = Ws2812Driver(116)

    @classmethod
    def tearDownClass(cls):
        """Tear down test fixtures, if any."""
        cls.pasek.close()

    def test_01_line_segment(self):


        self.pasek.reset_data()

        layer1 = self.pasek.add_segment(0,116,0)
        layer1.show_scale(10,110,70,100)

        layer2 = self.pasek.add_segment(0, 5, 100)
        layer2.show_animation(100)


















travis_pypi_setup.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
"""Update encrypted deploy password in Travis config file."""


from __future__ import print_function
import base64
import json
import os
from getpass import getpass
import yaml
from cryptography.hazmat.primitives.serialization import load_pem_public_key
from cryptography.hazmat.backends import default_backend
from cryptography.hazmat.primitives.asymmetric.padding import PKCS1v15


try:
    from urllib import urlopen
except ImportError:
    from urllib.request import urlopen


GITHUB_REPO = 'phephik/ws2812_driver'
TRAVIS_CONFIG_FILE = os.path.join(
    os.path.dirname(os.path.abspath(__file__)), '.travis.yml')


def load_key(pubkey):
    """Load public RSA key.

    Work around keys with incorrect header/footer format.

    Read more about RSA encryption with cryptography:
    https://cryptography.io/latest/hazmat/primitives/asymmetric/rsa/
    """
    try:
        return load_pem_public_key(pubkey.encode(), default_backend())
    except ValueError:
        # workaround for https://github.com/travis-ci/travis-api/issues/196
        pubkey = pubkey.replace('BEGIN RSA', 'BEGIN').replace('END RSA', 'END')
        return load_pem_public_key(pubkey.encode(), default_backend())


def encrypt(pubkey, password):
    """Encrypt password using given RSA public key and encode it with base64.

    The encrypted password can only be decrypted by someone with the
    private key (in this case, only Travis).
    """
    key = load_key(pubkey)
    encrypted_password = key.encrypt(password, PKCS1v15())
    return base64.b64encode(encrypted_password)


def fetch_public_key(repo):
    """Download RSA public key Travis will use for this repo.

    Travis API docs: http://docs.travis-ci.com/api/#repository-keys
    """
    keyurl = 'https://api.travis-ci.org/repos/{0}/key'.format(repo)
    data = json.loads(urlopen(keyurl).read().decode())
    if 'key' not in data:
        errmsg = "Could not find public key for repo: {}.\n".format(repo)
        errmsg += "Have you already added your GitHub repo to Travis?"
        raise ValueError(errmsg)
    return data['key']


def prepend_line(filepath, line):
    """Rewrite a file adding a line to its beginning."""
    with open(filepath) as f:
        lines = f.readlines()

    lines.insert(0, line)

    with open(filepath, 'w') as f:
        f.writelines(lines)


def load_yaml_config(filepath):
    """Load yaml config file at the given path."""
    with open(filepath) as f:
        return yaml.load(f)


def save_yaml_config(filepath, config):
    """Save yaml config file at the given path."""
    with open(filepath, 'w') as f:
        yaml.dump(config, f, default_flow_style=False)


def update_travis_deploy_password(encrypted_password):
    """Put `encrypted_password` into the deploy section of .travis.yml."""
    config = load_yaml_config(TRAVIS_CONFIG_FILE)

    config['deploy']['password'] = dict(secure=encrypted_password)

    save_yaml_config(TRAVIS_CONFIG_FILE, config)

    line = ('# This file was autogenerated and will overwrite'
            ' each time you run travis_pypi_setup.py\n')
    prepend_line(TRAVIS_CONFIG_FILE, line)


def main(args):
    """Add a PyPI password to .travis.yml so that Travis can deploy to PyPI.

    Fetch the Travis public key for the repo, and encrypt the PyPI password
    with it before adding, so that only Travis can decrypt and use the PyPI
    password.
    """
    public_key = fetch_public_key(args.repo)
    password = args.password or getpass('PyPI password: ')
    update_travis_deploy_password(encrypt(public_key, password.encode()))
    print("Wrote encrypted password to .travis.yml -- you're ready to deploy")


if '__main__' == __name__:
    import argparse
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument('--repo', default=GITHUB_REPO,
                        help='GitHub repo (default: %s)' % GITHUB_REPO)
    parser.add_argument('--password',
                        help='PyPI password (will prompt if not provided)')

    args = parser.parse_args()
    main(args)


ws2812_driver/ws2812_driver.py
from __future__ import print_function

import alsaaudio
import struct
import numpy as np
import time


class Ws2812Driver:

    _CONV_TABLE = None

    def __init__(self, leds, device='hw:CARD=snddaudio'):
        """

        :param leds: number of leds for driver
        :param device: refers to sound card
        Initialized control over sound card with fixed parametes
        Sample rate 88,2 kHz provides 0,3543us period of logic level.

        """

        self.num_leds = leds
        self._device = alsaaudio.PCM(device=device)
        self._device.setformat(alsaaudio.PCM_FORMAT_U16_BE)
        ch = 2
        sr = 88200
        period_size = 2048
        self._device.setchannels(ch)
        self._device.setrate(sr)
        self._device.setperiodsize(period_size)
        zeros = bytes(period_size * ch * 2 * [0])
        self._device.write(zeros)
        self._segments = []
        self.data = np.zeros((self.num_leds,3))
        self.overlay_matrix = np.ones((self.num_leds, 1))

        if not self._CONV_TABLE:
            self.__class__._CONV_TABLE = self._conv_table()

    def close(self):
        self._device.close()

    @classmethod
    def _conv_table(cls):
        """
        WS2812 works with period 1,23us±600ns
        Driver use 3x0,3543us long period to define logical value.
        Conversion table provides convert to hardware requirements:
        T0H --> 0 code, high voltage time   0,4us±150ns    --> 0 code
        T1H --> 1 code, high voltage time   0,8us±150ns    --> 11 code
        T0L --> 0 code, low voltage time    0.85us±150ns   --> 00 code
        T1L --> 0 code, low voltage time    0.45us±150ns   --> 0 code
        e.g. log. 1 --> T1H + T1L is coded as 110.
        """

        table = 256 * ['']

        for i in range(256):

            bin_string = ''
            for num in '{0:08b}'.format(i):
                add_string = '100' if num == '0' else '110'
                bin_string = bin_string + add_string
                print(bin_string)
                new_string = bin_string[16:] + bin_string[8:16] + bin_string[:8]
            table[i] = struct.pack('L', int(new_string, 2))[:-1]
            print(table[i])

        return table

    @classmethod
    def _print_nice(cls, bts):
        """
        It shows sent bits to device.
        """
        temp = ":".join(["{0:08b}".format(x) for x in bts])
        print(temp)

    @classmethod
    def _convert(cls, byte_val):
        """
        Select 24bit for one colour from convertion table
        """
        assert byte_val <= 256
        sel = cls._CONV_TABLE[byte_val]

        return sel

    @classmethod
    def _set_rgb(cls, red, green, blue):
        """
        Create GRB string of 72 bits for one led segment
        """
        r = cls._convert(int(red))
        g = cls._convert(int(green))
        b = cls._convert(int(blue))
        grb = g + r + b

        return grb

    def reset_data(self):

        data = bytes(((250 // 8) + 1) * [0])
        self._device.write(data)

    def frame_data(self, frame_buffer):
        """
        Data from frame_buffer is transformed to readable frame for sending to device.
        Each frame starts with RES low voltage time, which lasts above 50us.
        """

        data = bytes(((250 // 8) + 1) * [0])

        for item in range(len(frame_buffer)):
            row = frame_buffer[item]
            data = data + self._set_rgb(*row)

        end = self.num_leds - len(frame_buffer)
        data = data + end * self._set_rgb(0, 0, 0)
        periodsize = 2048
        prev_lenght = len(data)

        if prev_lenght % 2 != 0:
            data = data + bytes(1*[0])

        data = bytes([c for t in zip(data[1::2], data[::2]) for c in t])

        lenght = len(data)
        diff = periodsize * 4 - lenght
        data = data + bytes(diff * [0])

        self._device.write(data)

        return data

    def show_raw(self, pixels):
        """
        In case of any change is callng to redraw function.
        """

        self.data = pixels
        self.redraw()

    def show_scale(self, min_temp, max_temp, temp, bright= 100, heatmap=None,
                   reverse=None):

        """
        This function shows thermometer or any scale-based measurement
        Colour for each LED is calculated from defined heatmap for any number
        of LEDS and origin.
        In case of reversal mode is final data frame reversed.


        :param min_temp: Minimum temperature
        :param max_temp: Maximum temperature
        :param temp: Actual temperature(in range of previous parameters)
        :param bright: Allow to set brightness (0-100)
        :param heatmap: Used just in case own-defined heatmap
        :param reverse: Parameter 1 sets reverse scale
        :return:
        """
        num = self.num_leds
        frame_buffer = np.array([[0, 0, 0]])

        if heatmap is None:
            heatmap = [(0, 100, (0, 0, 255)),
                       (400, 500, (0, 255, 0)),
                       (900, 1000, (255, 0, 0))]

        br = bright / 100
        i = 0
        leds = (num / 1000)
        range_rel = max_temp - min_temp
        show_rel = temp - min_temp
        temp_abs = 1000 * (show_rel / range_rel)

        for area in heatmap:

            if i < (len(heatmap) - 1):
                next_heatmap = heatmap[i + 1]
                i = i + 1

            l_min = area[0]
            l_max = area[1]
            r1, g1, b1 = area[2]
            r1, g1, b1 = r1 * br, g1 * br, b1 * br
            step_size = next_heatmap[0] - l_max
            r2, g2, b2 = next_heatmap[2]
            r2, g2, b2 = r2 * br, g2 * br, b2 * br
            next_min = l_max + step_size
            step_leds = (next_min - l_max) * leds
            gap_r = (r2 - r1) / step_leds
            gap_g = (g2 - g1) / step_leds
            gap_b = (b2 - b1) / step_leds

            for led in range(int((l_min * leds)), int(l_max * leds)):

                if led < temp_abs * leds:
                    frame_buffer = np.concatenate((frame_buffer, [[r1, g1, b1]]))

                else:
                    frame_buffer = np.concatenate((frame_buffer, [[0, 0, 0]]))

            for led in range((int(l_max * leds)), (int(next_min * leds))):

                if led < temp_abs * leds:
                    r1, g1, b1 = r1 + gap_r, g1 + gap_g, b1 + gap_b
                    frame_buffer = np.concatenate((frame_buffer, [[r1, g1, b1]]))

                else:
                    frame_buffer = np.concatenate((frame_buffer, [[0, 0, 0]]))

        if reverse == 1:
            frame_buffer = frame_buffer[::-1]
        frame_buffer = np.delete(frame_buffer, 0, 0)
        frame_buffer = frame_buffer[:num, :]
        self.show_raw(frame_buffer)

    def add_segment(self, origin, num_leds, alfa):

        """
        Add a new segment
        :param origin: Set first LED on strip.
        :param num_leds: Set number of used LEDs
        :param alfa: Set transparency
        :return:
        """

        last_layer = self._segments[-1].layer if self._segments else 0
        temp = LineSegment(origin, num_leds, alfa, last_layer+1, self)
        self._segments.append(temp)
        self._update_overlay_matrix()

        return temp

    def _update_overlay_matrix(self):
        """
        Overlay matrix is used for set number of layer.
        """

        temp_matrix = np.ones((self.num_leds, 1))

        for segment in self._segments:

            overlay = temp_matrix[segment.origin:segment.origin + segment.num_leds]
            overlay += 1

        self.overlay_matrix = temp_matrix

    def redraw(self):
        """
        Redraw function provides for each pixels right drawing.
        In case of multiple layers are pixels averaging.
        """

        add_matrix = self.overlay_matrix
        temp = np.copy(self.data)

        for segment in self._segments:
            overlapped = temp[segment.origin:segment.origin + segment.num_leds]
            overlapped += segment.data * (1 - (segment.alfa / 255))

        for color in temp.T:
            color = temp/add_matrix

        temp = np.where(temp > 255, 255, temp)
        self.frame_data(temp)

    def show_animation(self, speed):
        """
        Show a fast trail on LED strip.
        Number of leds is parameter for how long will trail be.
        """
        speed = (1.1 - (speed/100))/10
        num = self.num_leds
        step_size = 255/num
        frame = np.zeros((num, 3))

        for m in range(num):
            frame[m, :] = 255, 255, 255
            pos = 255
            for o in range(m):
                pos -= step_size
                frame[m-o-1, :] = pos, pos, pos
        for p in range(num+1):
            zer = np.zeros((num-p, 3))
            frame1 = np.insert(zer, 0, frame, axis=0)
            frame2 = np.delete(frame1, range(0, num-p), axis=0)
            self.show_raw(frame2)
            time.sleep(speed)
        for n in range(116-num):
            self.origin += 1
            self.show_raw(frame)
            time.sleep(speed)
        for o in range(0, num-1):
            del_row = num-o-1
            self.origin += 1
            frame = np.delete(frame, del_row, axis=0)
            self.show_raw(frame)
            time.sleep(speed)

        frame = np.zeros((num, 3))
        self.origin = 1
        self.show_raw(frame)
        time.sleep(speed)


class LineSegment(Ws2812Driver):

    def __init__(self, origin, num_leds, alfa, layer, driver):
        """

        :param origin: Set first LED on strip
        :param num_leds: Set number of LEDs
        :param alfa: Set transparency
        :param layer: Set number of layers
        :param driver: Set used driver
        """
        self._num_leds = num_leds
        self._origin = origin
        self.alfa = alfa
        self.layer = layer
        self._driver = driver
        self.data = np.zeros(0)

    @property
    def num_leds(self):
        return self._num_leds

    @num_leds.setter
    def num_leds(self, value):
        self._num_leds = value
        self._driver._update_overlay_matrix()

    @property
    def origin(self):
        return self._origin

    @origin.setter
    def origin(self, value):
        self._origin = value
        self._driver._update_overlay_matrix()

    def show_raw(self, frame_data):
        self.data = frame_data
        self._driver.redraw()
